# 06 — Forecasting & Inventory Intelligence

 Notebook ini bertujuan untuk mengubah hasil demand forecasting menjadi aksi bisnis (*actionable reorder recommendation*). Alur proses yang dijalankan meliputi:

1. **Historical Sales & Inventory Ingestion**: Mengambil data stok harian dan histori transaksi demand.
2. **Missing Demand Imputation**: Membentuk grid data lengkap ($181 \text{ hari} \times 3 \text{ store} \times 80 \text{ produk} = 43,440 \text{ rows}$) untuk menghindari *missing record*.
3. **Daily Inventory Intelligence**: Menghitung *rolling average demand* (7d & 28d), *Days of Inventory*, dan *Inventory Risk Level*.
4. **Reorder Recommendation System**: Menghitung estimasi demand 7 hari ke depan, *safety stock* (20%), serta rekomendasi jumlah unit yang perlu dipesan kembali (*reorder quantity*).
5. **Executive Inventory Summary**: Ringkasan risiko persediaan pada posisi tanggal terbaru.

In [1]:
# Setup Library & Database Engine Connection
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
from IPython.display import display

# Tambahkan connect_timeout agar tidak hanging jika database bermasalah
engine = create_engine(
    "postgresql://postgres:admin@localhost:5432/smart_retail",
    connect_args={"connect_timeout": 5}
)

# Test koneksi secara eksplisit
try:
    with engine.connect() as conn:
        print("✅ Koneksi ke PostgreSQL Berhasil!")
except Exception as e:
    print("❌ Gagal terhubung ke database. Detail error:")
    print(e)

✅ Koneksi ke PostgreSQL Berhasil!


## Step 1: Ingestion Data Inventory & Sales Demand

Pada tahap ini, kita menarik data histori penutupan stok harian (*closing stock*) dan data total transaksi penjualan (*demand*) yang dikelompokkan per hari, toko, dan produk dari database.

In [2]:
# 1. Ambil data inventory (closing stock & stockout flag)
query_inventory = """
SELECT
    i.inventory_date,
    i.store_id,
    i.product_id,
    i.closing_stock,
    i.stockout_flag
FROM inventory i
ORDER BY
    i.inventory_date,
    i.store_id,
    i.product_id;
"""
inventory_df = pd.read_sql(
    query_inventory,
    engine,
    parse_dates=["inventory_date"]
)

# 2. Ambil data demand (sum transaction_qty per hari)
query_demand = """
SELECT
    transaction_date,
    store_id,
    product_id,
    SUM(transaction_qty) AS demand
FROM sales_transactions
GROUP BY
    transaction_date,
    store_id,
    product_id
ORDER BY
    transaction_date,
    store_id,
    product_id;
"""
demand_df = pd.read_sql(
    query_demand,
    engine,
    parse_dates=["transaction_date"]
)

# Verifikasi ringkas data yang telah ditarik
print("Inventory Data Sample:")
display(inventory_df.head())
print("\nDemand Data Sample:")
display(demand_df.head())

Inventory Data Sample:


,inventory_date,store_id,product_id,closing_stock,stockout_flag
0,2023-01-01,3,1,55,False
1,2023-01-01,3,2,128,False
2,2023-01-01,3,3,76,False
3,2023-01-01,3,4,79,False
4,2023-01-01,3,5,72,False



Demand Data Sample:


,transaction_date,store_id,product_id,demand
0,2023-01-01,3,22,5
1,2023-01-01,3,23,7
2,2023-01-01,3,24,5
3,2023-01-01,3,25,5
4,2023-01-01,3,26,4


## Step 2: Perbaiki Missing Demand (Full Grid Construction)

Hari tanpa penjualan sering kali tidak tercatat dalam tabel transaksi sales. Untuk menghitung *rolling average demand* secara akurat, kita wajib membuat kombinasi lengkap (*Cartesian product*) dari seluruh tanggal, toko, dan produk.

Target total baris:  
$$\text{181 hari} \times \text{3 stores} \times \text{80 products} = \mathbf{43,440\text{ rows}}$$

In [3]:
# Generate range tanggal lengkap dari min ke max transaction_date
dates = pd.date_range(
    demand_df["transaction_date"].min(),
    demand_df["transaction_date"].max(),
    freq="D"
)

# Ambil master store_id dan product_id
stores = pd.read_sql("SELECT store_id FROM stores", engine)
products = pd.read_sql("SELECT product_id FROM products", engine)

# Buat MultiIndex produk kartesian (dates x stores x products)
full_grid = (
    pd.MultiIndex.from_product(
        [dates, stores["store_id"], products["product_id"]],
        names=["transaction_date", "store_id", "product_id"]
    )
    .to_frame(index=False)
)

# Merge grid lengkap dengan demand_df & isi NaN demand dengan 0
full_demand = full_grid.merge(
    demand_df,
    on=["transaction_date", "store_id", "product_id"],
    how="left"
)
full_demand["demand"] = full_demand["demand"].fillna(0)

# Verifikasi jumlah baris yang terbentuk
print("Total Rows setelah full grid merge:", len(full_demand))

Total Rows setelah full grid merge: 43440


## Step 3: Penggabungan Data Inventory & Demand

Menggabungkan data grid demand lengkap dengan data histori *closing stock* harian menggunakan skema *left join* ber-validasi 1-to-1.

In [4]:
# Merge full_demand dengan inventory_df
inventory_daily = full_demand.merge(
    inventory_df,
    left_on=["transaction_date", "store_id", "product_id"],
    right_on=["inventory_date", "store_id", "product_id"],
    how="left",
    validate="one_to_one"
)

# Tampilkan jumlah baris dan beberapa baris pertama
print("Rows setelah merge inventory:", len(inventory_daily))
display(inventory_daily.head())

Rows setelah merge inventory: 43440


,transaction_date,store_id,product_id,demand,inventory_date,closing_stock,stockout_flag
0,2023-01-01,3,1,0.0,2023-01-01,55,False
1,2023-01-01,3,2,0.0,2023-01-01,128,False
2,2023-01-01,3,3,0.0,2023-01-01,76,False
3,2023-01-01,3,4,0.0,2023-01-01,79,False
4,2023-01-01,3,5,0.0,2023-01-01,72,False


## Step 4: Menghitung Rolling Demand Metrics (7d & 28d)

Menghitung rata-rata penjualan harian dalam rentang 7 hari (`avg_daily_demand_7d`) dan 28 hari (`avg_daily_demand_28d`). 

*Catatan:* Digunakan `.shift(1)` agar tidak terjadi *data leakage* (penjualan hari ini tidak boleh memprediksi demand hari ini sendiri).

In [5]:
# Urutkan data berdasarkan urutan waktu per store dan product
inventory_daily = inventory_daily.sort_values(
    ["store_id", "product_id", "transaction_date"]
)

# Grouping per (store_id, product_id) pada kolom demand
group = inventory_daily.groupby(["store_id", "product_id"])["demand"]

# Rolling mean 7 hari (diberi shift 1 agar murni historical)
inventory_daily["avg_daily_demand_7d"] = group.transform(
    lambda x: x.shift(1).rolling(7).mean()
)

# Rolling mean 28 hari
inventory_daily["avg_daily_demand_28d"] = group.transform(
    lambda x: x.shift(1).rolling(28).mean()
)

# Tampilkan sampel data hasil perhitungan rolling demand
display(inventory_daily[["transaction_date", "store_id", "product_id", "demand", "avg_daily_demand_7d", "avg_daily_demand_28d"]].tail())

,transaction_date,store_id,product_id,demand,avg_daily_demand_7d,avg_daily_demand_28d
42479,2023-06-26,8,87,18.0,18.000000,13.571429
42719,2023-06-27,8,87,24.0,18.285714,13.785714
42959,2023-06-28,8,87,11.0,20.857143,14.214286
43199,2023-06-29,8,87,19.0,18.857143,14.428571
43439,2023-06-30,8,87,12.0,19.142857,14.714286


## Step 5: Perhitungan Days of Inventory (DOI)

*Days of Inventory* menunjukkan berapa hari stok saat ini (`closing_stock`) akan bertahan berdasarkan laju konsumsi rata-rata 7 hari terakhir.
$$\text{Days of Inventory} = \frac{\text{Closing Stock}}{\text{Average Daily Demand (7d)}}$$

In [6]:
# Menghitung ketahanan stok dalam satuan hari
inventory_daily["days_of_inventory"] = (
    inventory_daily["closing_stock"] / inventory_daily["avg_daily_demand_7d"]
)

# Hindari nilai infinity akibat pembagian dengan nol (zero demand)
inventory_daily["days_of_inventory"] = inventory_daily["days_of_inventory"].replace(
    [np.inf, -np.inf], np.nan
)

# Cek distribusi nilai Days of Inventory
print("Ringkasan Statistik Days of Inventory:")
display(inventory_daily["days_of_inventory"].describe())

Ringkasan Statistik Days of Inventory:


count    38256.000000
mean        82.997659
std        162.602091
min          0.000000
25%          8.500000
50%         14.700000
75%         51.333333
max       1218.000000
Name: days_of_inventory, dtype: float64

## Step 6: Klasifikasi Risk Level Inventaris

Mengelompokkan stok produk ke dalam kategori risiko persediaan berdasarkan nilai *Days of Inventory*:
- **No Demand History**: Tidak ada histori demand (`NaN`).
- **Critical**: Ketahanan stok $\le 2$ hari.
- **High**: Ketahanan stok $2 - 5$ hari.
- **Medium**: Ketahanan stok $5 - 10$ hari.
- **Low**: Ketahanan stok $> 10$ hari.

In [7]:
# Fungsi pengkategorian risiko persediaan
def classify_inventory_risk(days):
    if pd.isna(days):
        return "No Demand History"
    if days <= 2:
        return "Critical"
    if days <= 5:
        return "High"
    if days <= 10:
        return "Medium"
    return "Low"

# Terapkan fungsi ke seluruh baris
inventory_daily["inventory_risk"] = inventory_daily["days_of_inventory"].apply(
    classify_inventory_risk
)

# Tampilkan jumlah kemunculan tiap risk level
print("Distribusi Inventory Risk Level (Seluruh Histori):")
print(inventory_daily["inventory_risk"].value_counts())

Distribusi Inventory Risk Level (Seluruh Histori):
inventory_risk
Low                  25886
Medium                9304
No Demand History     5184
High                  2815
Critical               251
Name: count, dtype: int64


## Step 7: Rekomendasi Reorder Quantity

Perhitungan kebutuhan restock dibuat menggunakan parameter:
1. **Forecast Demand 7-Hari**: $\text{avg\_daily\_demand\_7d} \times 7$
2. **Safety Stock (20%)**: $20\% \times \text{Forecast Demand 7d}$
3. **Recommended Stock Target**: $\text{Forecast Demand 7d} + \text{Safety Stock}$
4. **Reorder Quantity**: $(\text{Recommended Stock} - \text{Closing Stock})$, dipotong pada batas minimum 0 (`.clip(lower=0)`).

In [8]:
# 1. Proyeksi demand 7 hari ke depan
inventory_daily["forecast_7d_demand"] = (
    inventory_daily["avg_daily_demand_7d"] * 7
)

# 2. Perhitungan Safety Stock (buffer 20%)
inventory_daily["safety_stock"] = (
    inventory_daily["forecast_7d_demand"] * 0.20
)

# 3. Total recommended stock target
inventory_daily["recommended_stock"] = (
    inventory_daily["forecast_7d_demand"] + inventory_daily["safety_stock"]
)

# 4. Rekomendasi jumlah unit yang harus dipesan kembali (Reorder Quantity)
inventory_daily["reorder_quantity"] = (
    inventory_daily["recommended_stock"] - inventory_daily["closing_stock"]
).clip(lower=0)

# Tampilkan sampel hasil perhitungan reorder
display(
    inventory_daily[[
        "store_id", "product_id", "closing_stock", 
        "forecast_7d_demand", "safety_stock", "recommended_stock", "reorder_quantity"
    ]].tail()
)

,store_id,product_id,closing_stock,forecast_7d_demand,safety_stock,recommended_stock,reorder_quantity
42479,8,87,79,126.0,25.2,151.2,72.2
42719,8,87,77,128.0,25.6,153.6,76.6
42959,8,87,88,146.0,29.2,175.2,87.2
43199,8,87,124,132.0,26.4,158.4,34.4
43439,8,87,85,134.0,26.8,160.8,75.8


## Step 8: Daftar Produk Prioritas Restock Hari Ini (Latest Date)

Filter data pada tanggal paling baru (*latest date*) dalam dataset untuk mengambil daftar produk yang membutuhkan order ulang (`reorder_quantity > 0`), diurutkan dari kuantitas terbanyak.

In [9]:
# Ambil tanggal transaksi paling akhir
latest_date = inventory_daily["transaction_date"].max()
print(f"Laporan Reorder per Tanggal: {latest_date.strftime('%Y-%m-%d')}\n")

# Filter produk yang butuh di-reorder pada hari ini
reorder_today = (
    inventory_daily[inventory_daily["transaction_date"] == latest_date]
    .query("reorder_quantity > 0")
    .sort_values("reorder_quantity", ascending=False)
)

# Tampilkan Top 20 produk paling mendesak untuk di-restock
reorder_today_display = reorder_today[[
    "store_id",
    "product_id",
    "closing_stock",
    "avg_daily_demand_7d",
    "forecast_7d_demand",
    "days_of_inventory",
    "inventory_risk",
    "reorder_quantity"
]]

reorder_today_display.head(20)

Laporan Reorder per Tanggal: 2023-06-30



,store_id,product_id,closing_stock,avg_daily_demand_7d,forecast_7d_demand,days_of_inventory,inventory_risk,reorder_quantity
43244,3,45,24,16.571429,116.0,1.448276,Critical,115.2
43302,5,23,39,17.142857,120.0,2.275000,High,105.0
43235,3,36,28,15.142857,106.0,1.849057,Critical,99.2
43229,3,30,46,17.000000,119.0,2.705882,High,96.8
43324,5,45,34,15.000000,105.0,2.266667,High,92.0
43382,8,23,26,13.714286,96.0,1.895833,Critical,89.2
43359,5,87,82,19.857143,139.0,4.129496,High,84.8
43256,3,57,74,18.714286,131.0,3.954198,High,83.2
43312,5,33,30,13.428571,94.0,2.234043,High,82.8
43323,5,44,54,16.142857,113.0,3.345133,High,81.6


## Step 9: Executive Inventory Risk Summary

Ringkasan tingkat eksekutif pada posisi tanggal terakhir untuk memberikan gambaran umum status persediaan di seluruh toko dan produk.

In [10]:
# Ringkasan agregat persediaan berdasarkan Inventory Risk Level pada posisi tanggal terbaru
inventory_summary = (
    inventory_daily[inventory_daily["transaction_date"] == latest_date]
    .groupby("inventory_risk")
    .agg(
        products=("product_id", "count"),
        avg_stock=("closing_stock", "mean"),
        avg_days_inventory=("days_of_inventory", "mean"),
        total_reorder=("reorder_quantity", "sum")
    )
    .reset_index()
)

# Tampilkan tabel summary eksekutif
inventory_summary

,inventory_risk,products,avg_stock,avg_days_inventory,total_reorder
0,Critical,4,25.250000,1.760935,385.0
1,High,43,49.372093,3.668589,2755.0
2,Low,86,92.046512,165.041120,0.0
3,Medium,74,79.527027,7.310763,1119.4
4,No Demand History,33,87.454545,NaN,0.0


## Step 10: Executive Insights & Takeaways

### Data Analyst vs. Intelligence System View
- **Data Analyst View**: *"Produk A ter-jual total 1.500 unit selama 6 bulan terakhir."* (Deskriptif / Retrospektif).
- **Intelligence System View**: *"Produk A di Store X diperkirakan membutuhkan 180 unit dalam 7 hari ke depan. Stok saat ini sisa 90 unit (hanya cukup untuk 3,5 hari / High Risk). Rekomendasi reorder = 126 unit (termasuk 20% safety stock)."* (Preskriptif / Actionable).

### Key Answers dari Notebook Ini:
1. **Produk Prioritas Restock**: Dapat diidentifikasi secara presisi pada `reorder_today.head(20)` berdasarkan kuantitas deficit tertinggi.
2. **Recommended Reorder Quantity**: Disesuaikan secara dinamis dengan formula $\text{Demand Forecast 7d} + \text{Safety Stock (20\%)} - \text{Current Stock}$.
3. **Store & Risk Mapping**: Eksekutif dapat memantau `inventory_summary` untuk melihat persentase produk yang tergolong kategori *Critical* dan total modal/unit yang diperlukan untuk pengadaan stok ulang.